src/kernel/bpf/prog.rs

```diff
@@ -61,6 +70,90 @@ impl FromStr for BpfTag {
+#[derive(Debug, Default)]
+pub struct BpfInfoCache {
+    /// Our mapping from BPF tag to program information.
+    ///
+    /// Program information is reference counted, because it can be
+    /// shared with sub-programs.
+    cache: RefCell<HashMap<BpfTag, Rc<sys::bpf_prog_info>>>,
+}
+
+impl BpfInfoCache {
+    /// Look up BPF program information from the cache.
+    #[cfg_attr(feature = "tracing", crate::log::instrument(level = tracing::Level::TRACE, skip_all, fields(tag, result), err(level = tracing::Level::INFO)))]
+    fn lookup(&self, tag: BpfTag) -> Result<Option<sys::bpf_prog_info>> {
+        let mut cache = self.cache.borrow_mut();
+        if let Some(info) = cache.get(&tag) {
+            return Ok(Some(**info))
+        }
+
+        // If we didn't find the tag inside the cache we have to assume
+        // that our cache is outdated. Given how BPF program information
+        // retrieval works, we have to iterate over all programs, so
+        // clear the cache to remove potentially stale entries.
+        let () = cache.clear();
+
+        let mut found = None;
+        let mut next_prog_id = 0;
+
+        while let Ok(prog_id) = sys::bpf_prog_get_next_id(next_prog_id) {
+            let fd = sys::bpf_prog_get_fd_from_id(prog_id).with_context(|| {
+                format!("failed to retrieve BPF program file descriptor for program {prog_id}")
+            })?;
+
+            let mut info = sys::bpf_prog_info::default();
+            let () =
+                sys::bpf_prog_get_info_from_fd(fd.as_raw_fd(), &mut info).with_context(|| {
+                    format!("failed to retrieve BPF program information for program {prog_id}")
+                })?;
+
+            // We are going to need to retrieve additional information
+            // about all sub-programs (with different tags but captured
+            // by the same `bpf_prog_info`).
+            let mut prog_tags = Vec::<BpfTag>::with_capacity(info.nr_prog_tags as _);
+            // SAFETY: `BpfTag` is `repr(transparent)` and just eight bytes,
+            //         meaning it's valid for any bit pattern, so we can
+            //         adjust the vector's length to its capacity.
+            let () = unsafe { prog_tags.set_len(prog_tags.capacity()) };
+
+            let mut info = sys::bpf_prog_info {
+                nr_prog_tags: info.nr_prog_tags,
+                prog_tags: prog_tags.as_mut_ptr() as _,
+                ..Default::default()
+            };
+            let () =
+                sys::bpf_prog_get_info_from_fd(fd.as_raw_fd(), &mut info).with_context(|| {
+                    format!("failed to retrieve BPF program information for program {prog_id}")
+                })?;
+
+            let info = Rc::new(info);
+
+            // We need to map all the known sub-programs to the
+            // information that we are working with as well.
+            for prog_tag in iter::once(BpfTag::from(info.tag)).chain(prog_tags) {
```

Comment:
are you sure you need to chain info.tag with prog_tags? I think the very first tag in info.prog_tags is the same as info.tag, so you can probably simplify this a bit

Reply:
I am not sure about anything without documentation that states the intended behavior :-O

But from my brief experimentation, what you say does not always seem to be the case.

```patch
--- src/kernel/bpf/prog.rs
+++ src/kernel/bpf/prog.rs
@@ -148,9 +148,13 @@ impl BpfInfoCache {

             let info = Rc::new(info);

+            println!("--------------------------");
+            println!("INFO TAG: {:?}", BpfTag::from(info.tag));
+            println!("PROG TAGS: {:?}", prog_tags);
             // We need to map all the known sub-programs to the
             // information that we are working with as well.
             for prog_tag in iter::once(BpfTag::from(info.tag)).chain(prog_tags) {
+                println!("BPF TAG: {prog_tag}");
                 if found.is_none() && tag == prog_tag {
                     found = Some(*info);
                 }

```

```
--------------------------
INFO TAG: BpfTag([238, 14, 37, 60, 120, 153, 58, 36])
PROG TAGS: [BpfTag([238, 14, 37, 60, 120, 153, 58, 36])]
BPF TAG: ee0e253c78993a24
BPF TAG: ee0e253c78993a24
--------------------------
INFO TAG: BpfTag([164, 165, 223, 144, 55, 247, 205, 227])
PROG TAGS: [BpfTag([121, 62, 102, 12, 75, 84, 23, 167]), BpfTag([124, 123, 198, 106, 27, 126, 238, 251])]
BPF TAG: a4a5df9037f7cde3
BPF TAG: 793e660c4b5417a7
BPF TAG: 7c7bc66a1b7eeefb
```

So it checks out for the no-sub-program case but not once there are actual subprograms present, it seems...

The existing logic handles both cases just fine.

Reply:
discussed offline, seems like kernel is inconsistent, no harm doing it the way you do, let's keep it